src/d05.act

```diff
@@ -0,0 +1,133 @@
+import aoc
+import file
+import testing
+
+
+class Page(object):
+    number: str
+    predecessors: set[str]
+
+    def __init__(self, number, rules):
+        self.number = number
+        if number not in rules:
+            self.predecessors = set()
+        else:
+            self.predecessors = rules[number]
+        # Why can't I do this?
+        #self.predecessors = rules.get_def(number, set())
```

Comment:
I think it's clearer with the post-if style for these initial assignments, like so:
```suggestion
        self.predecessors = rules[number] if number in rules else set()
```

I'm actually a little surprised your code works because I'd like to remember that the compiler wasn't smart enough to follow all conditional branches to determine if non-optional attributes are all set. Maybe it's different for actors.

An alternative for more complex cases is to write a small helper function that returns non-optional...
```python
    self.predecessors = _def_predecessors()
```